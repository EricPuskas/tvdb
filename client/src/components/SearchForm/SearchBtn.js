import React from "react";

const SearchBtn = () => {
  const searchBtnTxt = window.innerWidth <= 600 ? null : "Search";
  return (
    <div className="input-group-append">
      <button
        className="btn btn-primary"
        type="submit"
        aria-label="Search Button"
      >
        <i className={"fas fa-search"} /> {searchBtnTxt}
      </button>
    </div>
  );
};

export default SearchBtn;
