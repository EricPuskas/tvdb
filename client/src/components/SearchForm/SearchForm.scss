form {
  width: 100%;
}

.Search-suggestions {
  font-size: 1.1rem;
  padding: 0.5rem;
  line-height: 1;
  border-left: 1px solid #555555;
  border-right: 1px solid #555555;
  &:first-child {
    border-top: 1px solid #555555;
  }
  &:last-child {
    border-bottom: 1px solid #555555;
  }

  &:hover {
    background-color: #a21616 !important;
    color: #fff;
    cursor: pointer;
  }
}

#__listItemProps {
  position: absolute;
  width: 100%;
  max-width: 800px;
  z-index: 999;
}

.form-control {
  height: calc(1.5em + 1rem + 2px);
  padding: 0.375rem 1.25rem;
  font-size: 1.3rem;
  border-radius: 4.25rem;
}

.form-control:focus {
  box-shadow: 0 0 0 0.2rem rgba(33, 34, 35, 0.25);
  border-color: #ffffff;
}

/* Extra small devices (phones, 600px and down) */
@media only screen and (max-width: 600px) {
  #__inputItemProps {
    height: calc(1.5em + 1rem + 2px);
    padding: 0.375rem 1.25rem;
    max-width: 95vw;
    font-size: 1rem;
  }
  #__listItemProps {
    position: relative;
    width: 100%;
    max-width: 100%;
    z-index: 999;
  }
  .Search-suggestions {
    font-size: 1rem;
  }
}
