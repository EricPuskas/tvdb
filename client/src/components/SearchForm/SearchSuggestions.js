import React from "react";
import uuid from "uuid/v4";

const SearchSuggestions = ({
  isOpen,
  loading,
  error,
  getItemProps,
  highlightedIndex,
  getListItemProps,
  items,
  selectedItem
}) => {
  const getSelectedStyles = (item, index) => ({
    backgroundColor:
      highlightedIndex === index ? "rgb(232, 232, 232)" : "#f9f9f9",
    fontWeight: selectedItem && selectedItem === item ? "bold" : "normal"
  });
  return (
    <>
      {((isOpen && !loading) || error) && (
        <div {...getListItemProps()} data-test="search-suggestions">
          {items.slice(0, 7).map((item, index) => (
            <div
              {...getItemProps({ item, index })}
              key={uuid()}
              className="Search-suggestions"
              style={getSelectedStyles(item, index)}
            >
              {item}
            </div>
          ))}
        </div>
      )}
    </>
  );
};

export default SearchSuggestions;
