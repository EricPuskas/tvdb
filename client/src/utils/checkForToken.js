import jwt_decode from "jwt-decode";

// Redux Store
import store from "../store";

// Utilities
import setToken from "./setToken";

// Actions
import { getToken } from "../actions/token";

const checkForToken = () => {
  // Check for token
  if (!localStorage.token) {
    store.dispatch(getToken());
    setToken(localStorage.token);
  } else {
    setToken(localStorage.token);
    // Decode token to be able to check if it expired
    const decoded = jwt_decode(localStorage.token);
    // Check for expired token
    const currentTime = Date.now() / 1000;
    if (decoded.exp < currentTime) {
      store.dispatch(getToken());
    }
  }
};

export default checkForToken;
