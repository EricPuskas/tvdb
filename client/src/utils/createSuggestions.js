const filterInput = input => {
  return input.replace(/[^\w\s]/gi, "");
};

const createSuggestions = (word, data) => {
  let filter = filterInput(word);
  const re = new RegExp(`${filter.toLowerCase()}.*\\B`, "g");
  return data.filter(item => re.test(item.toLowerCase()));
};

export default createSuggestions;
