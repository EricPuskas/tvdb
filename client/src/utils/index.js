import { applyMiddleware, createStore } from "redux";
import rootReducer from "../reducers";
import { middleware } from "../store";

export const findByAttribute = (component, attr) =>
  component.find(`[data-test="${attr}"]`);

export const testStore = initState => {
  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  return createStoreWithMiddleware(rootReducer, initState);
};
