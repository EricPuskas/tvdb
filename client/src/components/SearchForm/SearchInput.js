import React from "react";
import SearchBtn from "./SearchBtn";
import SearchSuggestions from "./SearchSuggestions";

const SearchInput = ({
  getInputProps,
  selectedItem,
  highlightedIndex,
  getItemProps,
  onSearchChange,
  searchText,
  isOpen,
  getListItemProps,
  items,
  loading
}) => {
  return (
    <>
      <div className="input-group mb-3">
        <input
          {...getInputProps({
            onChange: onSearchChange
          })}
          className="form-control"
          type="search"
          value={searchText}
          placeholder="Find your favorite shows"
          aria-label="Search Input"
        />
        <SearchBtn />
      </div>
      <SearchSuggestions
        loading={loading}
        searchText={searchText}
        getItemProps={getItemProps}
        selectedItem={selectedItem}
        highlightedIndex={highlightedIndex}
        isOpen={isOpen}
        getListItemProps={getListItemProps}
        items={items}
      />
    </>
  );
};

export default SearchInput;
