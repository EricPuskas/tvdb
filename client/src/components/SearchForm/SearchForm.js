import React, { useState } from "react";
import Suggestion from "search-suggestion";
import escapeRegex from "../../utils/escapeRegex";
import createSuggestions from "../../utils/createSuggestions";
import "./SearchInput";
import tvShows from "../../utils/tvShows";
import SearchInput from "./SearchInput";

const SearchForm = ({ history, searchText, setSearchText, loading }) => {
  const [suggestions, setSuggestions] = useState([]);

  const onSearchChange = e => {
    let value = e.target.value;
    let filterData = [];

    setSearchText(value);
    if (value) {
      filterData = createSuggestions(value, tvShows);
      setSuggestions(filterData);
    }
  };

  const handleSearchForm = e => {
    e.preventDefault();
    if (searchText && searchText !== "") {
      let search_query = escapeRegex(searchText);
      history.push(`/search?q=${search_query}`);
    }
    return;
  };

  const handleSearch = item => {
    let search_query;
    if (item) {
      search_query = escapeRegex(item);
      history.push(`/search?q=${search_query}`);
      setSearchText(search_query);
    } else if (searchText !== "") {
      search_query = escapeRegex(searchText);
      setSearchText(search_query);
      history.push(`/search?q=${search_query}`);
    }
  };

  const handleSuggestion = item => {
    setSearchText(item);
    handleSearch(item);
  };
  return (
    <form onSubmit={e => handleSearchForm(e)} autoComplete="off">
      <div className="form-group">
        <Suggestion
          items={suggestions}
          onSelectedItem={item => handleSuggestion(item)}
        >
          {props => (
            <SearchInput
              {...props}
              loading={loading}
              onSearchChange={onSearchChange}
              searchText={searchText}
            />
          )}
        </Suggestion>
      </div>
    </form>
  );
};

export default SearchForm;
