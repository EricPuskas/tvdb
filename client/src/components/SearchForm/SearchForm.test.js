import React from "react";
import { mount } from "enzyme";
import { findByAttribute } from "../../utils";
import SearchForm from "./SearchForm";

describe("SearchForm Component", () => {
  const component = mount(<SearchForm />);
  const wrapper = findByAttribute(component, "search-form");

  it("Should render without errors", () => {
    expect(wrapper.length).toBe(1);
  });

  it("Should pass snapshot test", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  const handleSubmit = jest.fn(e => jest.fn(e));
  const form = wrapper.find(`form`);
  form.simulate("submit");

  // Input is empty
  expect(handleSubmit).toHaveBeenCalledTimes(0);
});
