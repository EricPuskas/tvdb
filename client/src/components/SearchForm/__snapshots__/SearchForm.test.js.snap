// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm Component Should pass snapshot test 1`] = `"<form autocomplete=\\"off\\" data-test=\\"search-form\\"><div class=\\"form-group\\"><div id=\\"autoComplete\\"><div data-test=\\"search-input\\"><div class=\\"input-group mb-3\\"><input id=\\"__inputItemProps\\" placeholder=\\"Find your favorite shows\\" class=\\"form-control\\" type=\\"search\\" aria-label=\\"Search Input\\" value=\\"\\"><div class=\\"input-group-append\\" data-test=\\"search-btn\\"><button class=\\"btn btn-primary\\" type=\\"submit\\" aria-label=\\"Search Button\\"><i class=\\"fas fa-search\\"></i> Search</button></div></div></div></div></div></form>"`;
